docs: add CSI documentation (#11203)
hashicorp/vault: website/content/docs/platform/k8s/csi/index.mdx

---
layout: docs
page_title: Vault CSI Provider
sidebar_title: Vault CSI Provider
description: >-
  The Vault CSI Provider allows pods to consume Vault secrets using CSI volumes.
---

# Vault CSI Provider

The Vault CSI Provider allows pods to consume Vault secrets using 
[CSI Secrets Store](https://github.com/kubernetes-sigs/secrets-store-csi-driver) volumes.

~> The Vault CSI Provider requires the [CSI Secret Store](https://github.com/kubernetes-sigs/secrets-store-csi-driver) 
  Driver to be installed.

## Overview

> At a high level, the CSI Secret Store controller allows users to create `SecretProviderClass` objects. 

Reviewer: ```suggestion
At a high level, the CSI Secrets Store driver allows users to create `SecretProviderClass` objects. 
```

This object defines which secret provider to use and what secrets to retrieve. When pods requesting CSI volumes 

> are created in Kubernetes, the CSI Secret Store will send the request to the Vault CSI Provider when the provider 

Reviewer: ```suggestion
are created, the CSI Secrets Store driver will send the request to the Vault CSI Provider if the provider 
```

"in Kubernetes" is implied by pods.

---
layout: docs
page_title: Vault CSI Provider
sidebar_title: Vault CSI Provider
description: >-
  The Vault CSI Provider allows pods to consume Vault secrets using CSI Secret Store
  volumes.
---

# Vault CSI Provider

The Vault CSI Provider allows pods to consume Vault secrets using 

> [CSI Secret Store](https://github.com/kubernetes-sigs/secrets-store-csi-driver) volumes.

Reviewer: ```suggestion
[CSI Secrets Store](https://github.com/kubernetes-sigs/secrets-store-csi-driver) volumes.
```

~> The Vault CSI Provider requires the [CSI Secret Store](https://github.com/kubernetes-sigs/secrets-store-csi-driver) 
  Driver to be installed.

## Overview

At a high level, the CSI Secret Store controller allows users to create `SecretStoreClass` objects. 

> This object defines which secret provider to use and what secrets to retrieve. When the provider 

Reviewer: There is a sentence missing here. SPCs existing don't do anything on their own. The request is instigated by a pod mounting a CSI secrets store volume, which in turn specifies the SPC.

Author: Fixed 👍

used is `vault`, the Vault CSI Provider will receive the request, use the pods service account to retrieve 

> the specified secrets from Vault, and share them with the pod via a host path volume.

Reviewer: ```suggestion
the specified secrets from Vault, and mount them into the pod's CSI volume.
```

The host path volume is shared with the driver, not the requesting pod, and the secrets will probably soon be sent back to the driver over gRPC instead of written directly to the host path volume. So I think a small tweak here makes it a bit more accurate and saves us needing to update it again in the near future.

> The secret is retrieved from Vault and populated to the host path volume during the `ContainerCreation` phase. 
> This means that the secrets requested from Vault will be populated in the host path volume before containers 
> start running.

Reviewer: ```suggestion
The secret is retrieved from Vault and populated to the CSI secrets store volume during the `ContainerCreation` phase. 
This means that pods will be blocked from starting until the secrets have been read from Vault and written to the volume.
```

### Features

The following features are supported by the Vault CSI Provider:

* All Vault secret engines supported.
* Authenticatation using the requesting pod's service account.
* TLS/mTLS communciations with Vault.
* Rendering Vault secrets to files.
* Syncing secrets to Kubernetes secrets to be used as environment variables.
* Installation via [Vault Helm](/docs/platform/k8s/helm)

## Authenticating with Vault

The primary method of authentication with Vault when using the Vault CSI Provider 
is the service account attached to the pod. At this time no other authentication methods 
are supported.

> For Kubernetes authentication, the service account must be bound to a Vault role and a

Reviewer: Maybe include a docs link

Author: Done!

policy granting access to the secrets desired.

A service account must be present to use the Vault CSI Provider with the Kubernetes
authentication method. It is _not_ recommended to bind Vault roles to the default service
account provided to pods if no service account is defined.

## Secret Provider Class Example

The following is an example of a Secret Provider Class using the `vault` provider:

```yaml
---
apiVersion: secrets-store.csi.x-k8s.io/v1alpha1
kind: SecretProviderClass
metadata:
  name: vault-db-creds
spec:
  # Vault CSI Provider
  provider: vault
  parameters:
    # Vault role name to use during login
    roleName: "app"
    # Vault's hostname

>     vaultAddress: "https://vault.vault:8200"

Reviewer: ```suggestion
    vaultAddress: "https://vault:8200"
```

    # TLS CA certification for validation
    vaultCACertPath: "/vault/tls/ca.crt"
    objects: |
      - objectName: "dbUsername"
        secretPath: "database/creds/db-app"
        secretKey: "username"
      - objectName: "dbPassword"
        secretPath: "database/creds/db-app"
        secretKey: "password"
    # "objectName" is an alias used within the SecretProviderClass to reference 
    # that specific secret. This will also be the filename containing the secret.
    # "secretPath" is the path in Vault where the secret should be retrieved.
    # "secretKey" is the key within the Vault secret response to extract a value from.
```

~> Secret Provider Class is a namespaced object in Kuberenetes.

## Using Secret Provider Classes

> An application pod uses the example above Secret Provider Class by mounting it as a CSI volume:

Reviewer: ```suggestion
An application pod uses the example Secret Provider Class above by mounting it as a CSI volume:
```
